exec/exec.go

```diff
@@ -0,0 +1,99 @@
+package exec
+
+import (
+	"context"
+	"errors"
+	"io"
+	"os/exec"
+	"time"
+)
+
+// dcos-go/exec is a os/exec wrapper. It implements io.Reader and can be used to read both STDOUT and STDERR.
+//
+// Usage:
+// ce := exec.Run("bash", []string{"infinite.sh"}, time.Second*3)
+//
+// io.Copy(os.Stdout, ce)
+// err := <- ce.Done
+// if err != nil {
+// 	log.Error(err)
+// }
+
+var (
+	// ErrContextRequired is the error returned by `func run` if the context is empty.
+	ErrContextRequired = errors.New("context cannot be empty")
+)
+
+// CommandExecutor is a structure returned by exec.Run
+// Cancel can be used by a user to interrupt a command execution.
+// Done is a channel the user can read in order to retrieve execution status. Possible statuses:
+//  <nil> command executed successfully, returned 0 exit code
+//  <exit status N> where N is non 0 exit status.
+//  <context deadline exceeded> means timeout was reached and command was killed.
+//  <context canceled>  means that command was canceled by a user.
+type CommandExecutor struct {
+	Cancel context.CancelFunc
+	Done   chan error
+
+	done chan error
+	pipe *io.PipeReader
+}
+
+// Read implements the io.Reader.
+// CommandExecutor will read from stdout and stderr
+func (c CommandExecutor) Read(p []byte) (int, error) {
+	return c.pipe.Read(p)
+}
+
+// RunWithTimeout is a main command in a package. It executes the command passed by a user and returns `CommandExecutor`
+// structure. User is able to cancel the command by running `CommandExecutor.Cancel()` function, wait for
+// `timeout` or wait for command to execute. The error will be sent to CommandExecutor.Done channel
+func RunWithTimeout(command string, arg []string, timeout time.Duration) (CommandExecutor, error) {
```

Comment:
now .. this might be overkill, but what do you think about something like this? (pardon any coding errors ... doing this in a git PR isn't ideal)
```go
// Option is a functional option that configures a CommandExecutor
type Option func(*Context, *CommandExecutor) error

// Timeout configures a CommandExecutor with a working Cancel func and the given timeout
func Timeout(timeout time.Duration) Option {
  return func(ctx *Context, ce *CommandExecutor) error {
    if timeout <= 0 {
      return ErrInvalidTimeout
    }
    newContext, cancel := context.WithTimeout(*ctx, timeout)
    *ctx = newContext
    ce.cancel = cancel
    return nil
  }
}

// Run spawns the given command and returns a handle to the running process in the form
// of a CommandExecutor
// Usage:
//     cmd, err: = exec.Run("ls", []string{"-laF", "/etc"}, exec.Timeout(5 * time.Second))
func Run(command string, arg []string, options ...Option)(CommandExecutor, error) {
  // by default Cancel is spineless unless someone configures an option to enable it
  commandExecutor := CommandExecutor{ Cancel: func() {}, Done: ..., done: ... }
  ctx := context.Background()

  // apply options to command executor
  for _, opt := range options {
    if opt != nil {
      err := opt(&ctx, &commandExecutor)
      if err != nil {
        return CommandExecutor{}, err
      }
    }
  }

  // spawn the command
  ...
}
```